pkg/middleware/cached.go

```diff
@@ -103,3 +110,74 @@ func (ctx *CachedOptlyMiddleware) UserCtx(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r.WithContext(ctx))
 	})
 }
+
+// FeatureCtx extracts the featureKey URL param and adds an optimizelyconfig.OptimizelyFeature to the request context.
+// If no such feature exists in the current config, returns 404
+// Note: featureKey must be available as a URL param, and ClientCtx middleware must run prior to this middleware
+func (mw *CachedOptlyMiddleware) FeatureCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		optlyClient, err := GetOptlyClient(r)
+		if err != nil {
+			RenderError(fmt.Errorf("optlyClient not available in FeatureCtx"), http.StatusInternalServerError, w, r)
+			return
+		}
+
+		featureKey := chi.URLParam(r, "featureKey")
+		if featureKey == "" {
+			RenderError(fmt.Errorf("invalid request, missing featureKey in FeatureCtx"), http.StatusBadRequest, w, r)
+			return
+		}
+
+		feature, err := optlyClient.GetFeature(featureKey)
+		var statusCode int
+		switch {
+		case err == nil:
+			GetLogger(r).Debug().Str("featureKey", featureKey).Msg("Added feature to request context in FeatureCtx")
+			ctx := context.WithValue(r.Context(), OptlyFeatureKey, &feature)
+			next.ServeHTTP(w, r.WithContext(ctx))
+			return
+		case errors.Is(err, optimizely.ErrEntityNotFound):
+			statusCode = http.StatusNotFound
+		default:
+			statusCode = http.StatusInternalServerError
+		}
+		GetLogger(r).Debug().Err(err).Str("featureKey", featureKey).Msg("Calling GetFeature in FeatureCtx")
+		RenderError(err, statusCode, w, r)
+	})
+}
+
+// ExperimentCtx extracts the experimentKey URL param and adds a optimizelyconfig.OptimizelyExperiment to the request context.
+// If no such experiment exists in the current config, returns 404
+// Note: experimentKey must be available as a URL param, and ClientCtx middleware must run prior to this middleware
+func (mw *CachedOptlyMiddleware) ExperimentCtx(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		optlyClient, err := GetOptlyClient(r)
+		if err != nil {
+			RenderError(fmt.Errorf("optlyClient not available in ExperimentCtx"), http.StatusInternalServerError, w, r)
+			return
+		}
+
+		experimentKey := chi.URLParam(r, "experimentKey")
+		if experimentKey == "" {
+			RenderError(fmt.Errorf("invalid request, missing experimentKey in ExperimentCtx"), http.StatusBadRequest, w, r)
+			return
+		}
+
+		experiment, err := optlyClient.GetExperiment(experimentKey)
+		var statusCode int
+		switch {
+		case err == nil:
+			GetLogger(r).Debug().Str("experimentKey", experimentKey).Msg("Added experiment to request context in ExperimentCtx")
+			ctx := context.WithValue(r.Context(), OptlyExperimentKey, &experiment)
+			next.ServeHTTP(w, r.WithContext(ctx))
+			return
+		case errors.Is(err, optimizely.ErrEntityNotFound):
+			statusCode = http.StatusNotFound
+		default:
+			statusCode = http.StatusInternalServerError
+		}
```

Comment:
Nice use of error handling. I'm not sure if this is the first time, but this does limit Agent to Go version 1.13. Something to consider.

Reply:
The only drawback I can think of is if we want to allow consuming Agent as a library, then the users need to be on 1.13.